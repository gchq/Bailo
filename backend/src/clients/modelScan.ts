import fetch, { Response } from 'node-fetch'
import { Readable } from 'stream'

import config from '../utils/config.js'
import { BadReq, InternalError } from '../utils/error.js'

interface ModelScanInfoResponse {
  apiName: string
  apiVersion: string
  scannerName: string
  modelscanVersion: string
}

interface ModelScanResponse {
  summary: {
    total_issues: number
    total_issues_by_severity: {
      LOW: number
      MEDIUM: number
      HIGH: number
      CRITICAL: number
    }
    input_path: string
    absolute_path: string
    modelscan_version: string
    timestamp: string
    scanned: {
      total_scanned: number
      scanned_files: string[]
    }
    skipped: {
      total_skipped: number
      skipped_files: string[]
    }
  }
  issues: {
    description: string
    operator: string
    module: string
    source: string
    scanner: string
    severity: string
  }[]
  errors: {
    category: string
    description: string
    source: string
  }[]
}

export async function getModelScanInfo() {
  const url = `${config.avScanning.modelscan.protocol}://${config.avScanning.modelscan.host}:${config.avScanning.modelscan.port}`
  let res: Response

  try {
    res = await fetch(`${url}/info`, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    })
  } catch (err) {
    throw InternalError('Unable to communicate with the ModelScan service.', { err })
  }
  if (!res.ok) {
    throw BadReq('Unrecognised response returned by the ModelScan service.')
  }

  return (await res.json()) as ModelScanInfoResponse
}

export async function scanStream(stream: Readable, fileName: string, fileSize: number) {
  const url = `${config.avScanning.modelscan.protocol}://${config.avScanning.modelscan.host}:${config.avScanning.modelscan.port}`
  let res: Response

  try {
    const formData = new FormData()
    formData.append(
      'in_file',
      {
        [Symbol.toStringTag]: 'File',
        size: fileSize,
        stream: () => stream,
      },
      fileName,
    )

    res = await fetch(`${url}/scan/file`, {
      method: 'POST',
      headers: {
        accept: 'application/json',
      },
      body: formData,
    })
  } catch (err) {
    throw InternalError('Unable to communicate with the ModelScan service.', { err })
  }
  if (!res.ok) {
    throw BadReq('Unrecognised response returned by the ModelScan service.', {
      body: JSON.stringify(await res.json()),
    })
  }

  return (await res.json()) as ModelScanResponse
}
