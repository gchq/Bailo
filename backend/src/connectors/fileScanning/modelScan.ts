import { Readable } from 'stream'

import { getModelScanInfo, scanStream } from '../../clients/modelScan.js'
import { getObjectStream } from '../../clients/s3.js'
import { FileInterfaceDoc } from '../../models/File.js'
import log from '../../services/log.js'
import config from '../../utils/config.js'
import { ConfigurationError } from '../../utils/error.js'
import { BaseFileScanningConnector, FileScanResult, ScanState } from './Base.js'

export const modelScanToolName = 'ModelScan'

export class ModelScanFileScanningConnector extends BaseFileScanningConnector {
  constructor() {
    super()
  }

  info() {
    return [modelScanToolName]
  }

  async init(retryCount: number = 1) {
    log.info('Initialising ModelScan...')
    if (retryCount <= config.connectors.fileScanners.maxInitRetries) {
      setTimeout(async () => {
        try {
          await getModelScanInfo()
          log.info('ModelScan initialised.')
        } catch (error) {
          log.warn(`Could not initialise ModelScan, retrying (attempt ${retryCount})...`)
          this.init(++retryCount)
        }
      }, config.connectors.fileScanners.initRetryDelay)
    } else {
      throw ConfigurationError(
        `Could not initialise Model Scan after ${retryCount} attempts, make sure that it is setup and configured correctly.`,
        {
          modelScanConfig: config.avScanning.modelscan,
        },
      )
    }
  }

  async scan(file: FileInterfaceDoc): Promise<FileScanResult[]> {
    this.init()

    let modelscanVersion: string | undefined = undefined
    try {
      modelscanVersion = (await getModelScanInfo()).modelscanVersion
    } catch (error) {
      log.error('Could not run ModelScan as it is not running', error)
      return [
        {
          toolName: modelScanToolName,
          scannerVersion: 'Unknown',
          state: ScanState.Error,
          lastRunAt: new Date(),
        },
      ]
    }

    const s3Stream = (await getObjectStream(file.bucket, file.path)).Body as Readable
    try {
      const scanResults = await scanStream(s3Stream, file.name, file.size)

      if (scanResults.errors.length !== 0) {
        log.error(
          { errors: scanResults.errors, modelId: file.modelId, fileId: file._id, name: file.name },
          'Scan errored.',
        )
        return [
          {
            toolName: modelScanToolName,
            state: ScanState.Error,
            scannerVersion: modelscanVersion,
            lastRunAt: new Date(),
          },
        ]
      }

      const issues = scanResults.summary.total_issues
      const isInfected = issues > 0
      const viruses: string[] = []
      if (isInfected) {
        for (const issue of scanResults.issues) {
          viruses.push(`${issue.severity}: ${issue.description}. ${issue.scanner}`)
        }
      }
      log.info(
        { modelId: file.modelId, fileId: file._id, name: file.name, result: { isInfected, viruses } },
        'Scan complete.',
      )
      return [
        {
          toolName: modelScanToolName,
          state: ScanState.Complete,
          scannerVersion: modelscanVersion,
          isInfected,
          viruses,
          lastRunAt: new Date(),
        },
      ]
    } catch (error) {
      log.error({ error, modelId: file.modelId, fileId: file._id, name: file.name }, 'Scan errored.')
      return [
        {
          toolName: modelScanToolName,
          state: ScanState.Error,
          scannerVersion: modelscanVersion,
          lastRunAt: new Date(),
        },
      ]
    }
  }
}
