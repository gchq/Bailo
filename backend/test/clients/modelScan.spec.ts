import { PassThrough } from 'stream'
import { describe, expect, test, vi } from 'vitest'

import { getModelScanInfo, scanStream } from '../../src/clients/modelScan.js'

const configMock = vi.hoisted(() => ({
  avScanning: {
    modelscan: {
      enabled: true,
      connection: {
        host: 'example.com',
      },
    },
  },
}))

vi.mock('../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMock,
}))

const fetchMock = vi.hoisted(() => ({
  default: vi.fn(() => ({ ok: true, text: vi.fn(), json: vi.fn() })),
}))
vi.mock('node-fetch', async () => fetchMock)

describe('clients > modelScan', () => {
  test('getModelScanInfo > success', async () => {
    const expectedResponse = {
      apiName: 'Bailo ModelScan API',
      apiVersion: '1.0.0',
      scannerName: 'modelscan',
      modelscanVersion: '0.8.1',
    }
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => expectedResponse),
    })
    const response = await getModelScanInfo()

    expect(fetchMock.default).toBeCalled()
    expect(fetchMock.default.mock.calls).toMatchSnapshot()
    expect(response).toStrictEqual(expectedResponse)
  })

  test('getModelScanInfo > bad response', async () => {
    fetchMock.default.mockResolvedValueOnce({
      ok: false,
      text: vi.fn(() => 'Unrecognised response'),
      json: vi.fn(),
    })

    expect(() => getModelScanInfo()).rejects.toThrowError(/^Unrecognised response returned by the ModelScan service./)
  })

  test('getModelScanInfo > rejected', async () => {
    fetchMock.default.mockRejectedValueOnce('Unable to communicate with the inferencing service.')

    expect(() => getModelScanInfo()).rejects.toThrowError(/^Unable to communicate with the ModelScan service./)
  })

  test('scanStream > success', async () => {
    const expectedResponse = {
      summary: {
        total_issues: 0,
        total_issues_by_severity: {
          LOW: 0,
          MEDIUM: 0,
          HIGH: 0,
          CRITICAL: 0,
        },
        input_path: '/foo/bar/safe_model.h5',
        absolute_path: '/foo/bar',
        modelscan_version: '0.8.1',
        timestamp: '2024-11-27T12:00:00.000000',
        scanned: {
          total_scanned: 1,
          scanned_files: ['safe_model.h5'],
        },
        skipped: {
          total_skipped: 0,
          skipped_files: [],
        },
      },
      issues: [],
      errors: [],
    }
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => expectedResponse),
    })
    // force lastModified to be 0
    const date = new Date(1970, 0, 1, 0)
    vi.setSystemTime(date)

    const response = await scanStream(new PassThrough(), 'safe_model.h5', 0)

    expect(fetchMock.default).toBeCalled()
    expect(fetchMock.default.mock.calls).toMatchSnapshot()
    expect(response).toStrictEqual(expectedResponse)
  })

  test('scanStream > bad response', async () => {
    fetchMock.default.mockResolvedValueOnce({
      ok: false,
      text: vi.fn(() => 'Unrecognised response'),
      json: vi.fn(),
    })
    expect(() => scanStream(new PassThrough(), 'safe_model.h5', 0)).rejects.toThrowError(
      /^Unrecognised response returned by the ModelScan service./,
    )
  })

  test('scanStream > rejected', async () => {
    fetchMock.default.mockRejectedValueOnce('Unable to communicate with the ModelScan service.')

    expect(() => scanStream(new PassThrough(), 'safe_model.h5', 0)).rejects.toThrowError(
      /^Unable to communicate with the ModelScan service./,
    )
  })
})
