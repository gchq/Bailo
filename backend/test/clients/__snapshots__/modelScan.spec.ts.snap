// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`clients > modelScan > getModelScanInfo > success 1`] = `
[
  [
    "undefined://undefined:undefined/info",
    {
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`clients > modelScan > scanStream > success 1`] = `
[
  [
    "undefined://undefined:undefined/scan/file",
    {
      "body": FormData {
        Symbol(state): [
          {
            "name": "in_file",
            "value": FileLike {
              Symbol(state): {
                "blobLike": {
                  "size": 0,
                  "stream": [Function],
                  Symbol(Symbol.toStringTag): "File",
                },
                "lastModified": 0,
                "name": "safe_model.h5",
                "type": undefined,
              },
            },
          },
        ],
      },
      "headers": {
        "accept": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;
